quarterly_stats Fixes: https://github.com/gitcoinco/web/issues/946

CuriousLearner opened the pull request:
<!--
  Thank you for your pull request. Please provide a description above and review
  the requirements below.

  Contributors guide: https://github.com/gitcoinco/web/blob/contributing/CONTRIBUTING.md
-->

##### Description
This is a work in progress for the new feature of sending quarterly status emails to users.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] linter status: 100% pass
- [x] changes don't break existing behavior
- [x] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, ui, crypto, etc). -->

##### Testing
<!-- Why should the PR reviewer trust that this change doesn't break anything? How have you tested this change? -->
Tests to follow

##### Refers/Fixes
Fixes: #946

owocki commented:
wooo hoo! looking forward to reviewing this when its ready

thelostone-mc commented on app/retail/templates/emails/quarterly_stats.html:
```diff
@@ -0,0 +1,72 @@
+        {% comment %}
+        TODO: Find where we can fetch the below info from.
+        {% endcomment %}
+        A total of value of $132,810 USD was transacted on the Gitcoin network-812 ETH, 20,000 DAI, and 12,400 WYV.
```
We don't have this yet! Leave it hardcoded , we'll add the integration 
Thoughts @mbeacom  ?

owocki commented:
lookin now

owocki commented:
pushed up a few commits to help with testing

just tested.. here is a screenshot of the email. 

![screencapture-localhost-8000-_administration-email-quarterly_roundup-2018-04-30-13_35_39](https://user-images.githubusercontent.com/513929/39446352-722e6674-4c7b-11e8-83df-9b9fbed7c0e0.png)

some notes:

* a few spacing issues
* a few of the stats look incorrect
* a few rounding errors (eth ETH amount needs to be divided by 10**18)
* as noted above, the images from the design need to be added

thelostone-mc commented:
@CuriousLearner So these were a few things I could pick out via the screenshot @owocki had put up

To add to his comments: 

- Make the width of the container thinner (like in the design )
- use font "Muli" + "Muli Bold" (it's there in our code base)
- the heading need to capitalized and not all uppercase 
- subheading needs to be greyed out
- give some breathing space between the content + badge (margin ? )
- the button texts and text in footer need not be bold 
- In the footer, breathing space between the logos

CuriousLearner commented:
Hi,

I've got a few questions:

1) How do I get GitHub profile image of someone to display?
2) What URL should be filled up with the two buttons namely: `Share my stats` & `Explore new bounties`
3) For the percentage part, I'm not sure how to do it in CSS. Can you help?

BDW, here is how it looks now.

@owocki  @thelostone-mc Addressed your reviews :)

![image](https://user-images.githubusercontent.com/8039608/39480705-a51f3ef8-4d86-11e8-84ca-5301e795d2db.png)

CuriousLearner commented:
Oh, bdw, that ether amount fix is also done.

thelostone-mc commented:
Explore new bounties button should redirect them to the explorer with the work open filter selected + network as mainnet.
All other filters should be removed 

For languages , we can display just 3 of them on the left side ?
Have 3 stacks which will be populated with 3 languages , the rest we can list in text 
Need to check with @PixelantDesign though ! What do you think of this ? 


Share button , should it be a tweet ? @owocki

The GitHub images as discussed offline , you can check out the profile / leaderboard `gitcoinco/static/avatar/<username>`

The percentage css : could you pull it using 
https://codepen.io/ianchen0419/pen/JRKbPy as reference ? ( Cant open my laptop cause I'm traveling back home in bus and my co passenger is forcing me into having a conversation  -.- ) If still stuck , I can help out later after I get back

CuriousLearner commented:
Hey, @thelostone-mc As I can see, it is and AJAX call with query params that fetch open filter selected bugs on mainnet. The URL remains same: `https://gitcoin.co/explorer`

So, should we keep it as https://gitcoin.co/explorer or is there some other way, we can handle this?

thelostone-mc commented:
Ah yeah ! Those are set via localStorage -.-
Yeah let's keep it as good ol explorer

thelostone-mc commented on app/retail/templates/emails/quarterly_stats.html:
```diff
@@ -0,0 +1,230 @@
+  .button-action {
+    margin-bottom: 5em;
+    margin-top: 3em;
+  }
```
Mind leaving a line after each CSS class ? (we stick to that throughout the repo)

thelostone-mc commented:
@CuriousLearner could you put up a screenshot -> so that we can get the design approval from @PixelantDesign ?

@mbeacom looked at the code, looks alright to me other than some indention issues 

@CuriousLearner will pull this in locally and check it out

CuriousLearner commented:
@thelostone-mc Addressed your review and here is the screenshot of current design:



![gitcoin](https://user-images.githubusercontent.com/8039608/39527937-6154ed66-4e40-11e8-9ab8-d1c470afde81.png)

thelostone-mc commented:
Yup! looks neat ! ❤️ 

Left a few comments and with that we should be good to go 

Last bit - 
- The bit where we've got 2 circles , the spacing looks off. We could either make the circles smaller or increase the space of the box 
- Give the section left and right padding (aka for every black box )
- Add some padding-top and bottom for the buttons
- For the circle % could you default it to  a circle with a thick border ? (aka what it would be if it was a 100%)
- Since rendering it dynamically is a pain -> could you use [this](https://fontawesome.com/icons/seedling?style=solid) in green as a temp placeholder
 (if you've a better one in mind, we can go with that )

And finally if you think it looks better : 
- vertical align the text ? 

Sorry , should have pointed this out earlier :/

CuriousLearner commented on app/dashboard/models.py:
```diff
@@ -61,7 +61,7 @@ def current(self):
     def stats_eligible(self):
         """Exclude results that we don't want to track in statistics."""
-        return self.exclude(idx_status__in=['unknown', 'cancelled'])
+        return self.exclude(current_bounty=True, idx_status__in=['unknown', 'cancelled'])
```
Ah, @owocki added this bit, I guess. So, I'm not sure about it.

@owocki can you please confirm this?

owocki replied:
confirmed.

thelostone-mc commented:
This looks neat ^_^

CuriousLearner commented:
@mbeacom @owocki Reviews please :)

owocki commented on app/dashboard/models.py:
```diff
@@ -1048,6 +1061,51 @@ def stats(self):
+        user_active_in_last_quarter = False
+        last_quarter = datetime.now() - timedelta(days=90)
+        bounties = self.bounties.filter(modified_on__gte=last_quarter)
+        loyalty_rate = 0
```
looks like this is never computed?

CuriousLearner replied:
This is a part of `stats` property not `quarterly_stats` property.

I just fixed some linting errors in the `stats` property.

CuriousLearner replied:
Oops! Yes, you were right on this. Removed that bit out.

owocki commented on scripts/crontab:
```diff
@@ -37,6 +37,7 @@ PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/us
 */15 * * * * cd gitcoin/coin; bash scripts/run_management_command_if_not_already_running.bash get_notifications  >> /var/log/gitcoin/get_notifications.log  2>&1
 15 10 * * * cd gitcoin/coin; bash scripts/run_management_command_if_not_already_running.bash bounty_feedback_email  >> /var/log/gitcoin/bounty_feedback_email.log  2>&1
 15 11 * * * cd gitcoin/coin; bash scripts/run_management_command_if_not_already_running.bash new_bounties_email  >> /var/log/gitcoin/new_bounties_email.log  2>&1
+0 0 1 */3 * cd gitcoin/coin; bash scripts/run_management_command_if_not_already_running.bash send_quarterly_stats  >> /var/log/gitcoin/send_quarterly_stats.log  2>&1
```
we will want to QA this before we send the first one :)

owocki replied:
also, you should add the `--live` param

owocki commented on app/marketing/mails.py:
```diff
@@ -323,6 +323,32 @@ def new_match(to_emails, bounty, github_username):
+            print("-----" * 100)
+            from_email = settings.PERSONAL_CONTACT_EMAIL
+
+            if not should_suppress_notification_email(to_email, 'transactional'):
```
can you change `transactional` to `roundup`

owocki commented:
qa notes

1. pls resolve merge conflicts
2. [rounding errors](http://bits.owocki.com/2h2Q2m470Y1z/Screen%20Shot%202018-05-07%20at%209.46.45%20AM.png)
3. [empty circle!](http://bits.owocki.com/0W3c1i3x0P2p/Screen%20Shot%202018-05-07%20at%209.47.25%20AM.png)
4. 'share my stats' page doesnt go anywhere
5. [mobile view of the email needs some work](http://bits.owocki.com/3Y352X26222N/Screen%20Shot%202018-05-07%20at%209.48.01%20AM.png)
6. [this number seems wrong](http://bits.owocki.com/1m352g2W152D/Screen%20Shot%202018-05-07%20at%209.48.32%20AM.png)

looking good so far, i just think we need to update those above issues!

thelostone-mc commented:
@CuriousLearner could we squash the commits ? ^_^

CuriousLearner commented:
Hi @owocki , please find my updates here:

1) Merge conflicts resolved
2) Rounding errors fixed
3) On it.
4) I'm still waiting to know where the `share my stats` page should go?
5) On it.
6) This is hard-coded for now since we didn't decide how do reach these stats.

looking good so far, i just think we need to update those above issues!

owocki commented:
> I'm still waiting to know where the share my stats page should go?

for now, i suppose let's just change the button to say "View Bounties" and have it go to gitcoin.co/explorer

owocki commented on app/retail/templates/emails/quarterly_stats.html:
```diff
@@ -0,0 +1,263 @@
+                {% comment %}
+                TODO: Find where we can fetch the below info from.
+                {% endcomment %}
+                <p>You've boosted your skills in 4 different languages: HTML, CSS, Python, and JavaScript!</p>
```
can you put these variables into the python layer as an hardcoded array and @mbeacom and i can fill them out with information from the backend?

owocki commented on app/marketing/utils.py:
```diff
@@ -141,3 +142,59 @@ def get_or_save_email_subscriber(email, source, send_slack_invite=True, profile=
+    hunters = [username[0] for username in profiles]
+
+    # Overall transactions across the network are hard-coded for now
+    total_transaction_in_usd = "132,810"
```
TODO: fill these in on the backend

owocki commented:
1. pulled down the latest, but [still seeing a rounding error here](http://bits.owocki.com/1R2G1P1G0a1f/Screen%20Shot%202018-05-08%20at%204.28.36%20PM.png)
1. can we make the [language image](http://bits.owocki.com/2Y1S0831011x/Screen%20Shot%202018-05-08%20at%204.29.58%20PM.png) look like the design?  should be do-able with some HTML/CSS fu
1. mobile view still looks funny ( i think you said you were on it though :) ) 
1. any way the percentages can look [more like the design](http://bits.owocki.com/1c1F350S2226/Screen%20Shot%202018-05-08%20at%204.31.08%20PM.png)?

looks like were getting close!  excited for this to get merged

CuriousLearner commented:
@owocki @thelostone-mc @mbeacom 


1) Rounding error is fixed.
2) Language image is added.
3) while working on mobile view, it seems like all the style descriptors are made inline with elements (may be because of some post-processor?), so the media queries aren't able to override the design. How should I go about this?
4) I've added the color to the percentages, but for filling in the circle with appropriate percentage, I've tried that already, but seemed complicated.

Let me know what I can do for the 3rd part and I can make this mobile responsive.

thelostone-mc commented on app/dashboard/models.py:
```diff
@@ -1065,6 +1078,60 @@ def stats(self):
+        total_bounties = bounties.count()
+        completetion_percent = int(
+            round(num_completed_bounties * 1.0 / total_bounties, 2) * 100
+        ) if total_bounties != 0 else '0'
```
This needs to be updated to `if total_bounties != 0 else 0`
Else it would throw an error below at 1113

thelostone-mc commented on app/dashboard/models.py:
```diff
@@ -1065,6 +1078,60 @@ def stats(self):
+        ) if total_bounties != 0 else '0'
+
+        avg_eth_earned_per_bounty = total_earned_eth / fulfilled_bounties_count
+        avg_usd_earned_per_bounty = total_earned_usd / fulfilled_bounties_count
```
worth checking if `fulfilled_bounties_count` is not 0 ?

thelostone-mc commented:
> while working on mobile view, it seems like all the style descriptors are made inline with elements (may be because of some post-processor?), so the media queries aren't able to override the design. How should I go about this?

@mbeacom Did we add anything new ? I pulled in the code and checked it out -> all the css classes get added inline which prevents from overwriting it via media queries

@SaptakS  any ideas ?

CuriousLearner commented:
@thelostone-mc Addressed your review comments. Not sure how to make media queries run when every style is applied inline :(

owocki commented:
> while working on mobile view, it seems like all the style descriptors are made inline with elements (may be because of some post-processor?), so the media queries aren't able to override the design. How should I go about this?

anyone have any thoughts here? the reason we use the preprocessor is that some email clients won't respect CSS elements unless the attributes are inline.

theres got to be some standard way of doing this

owocki commented:
> @thelostone-mc Addressed your review comments. Not sure how to make media queries run when every style is applied inline :(

now that i think about this.. maybe we should just be building the email as 'mobile first' so that it looks good in mobile and in desktop with the same css

owocki commented:
> I've added the color to the percentages, but for filling in the circle with appropriate percentage, I've tried that already, but seemed complicated.

maybe instead of doing a circle we just something similar like a `<span>` with a partial fill inside of it

something like this:

```
<span style="color:grey">
<span style="width: 75%; color:black">&nbsp;</span>
</span>
```

owocki commented:
follow up tickett https://github.com/gitcoinco/web/issues/1179

owocki commented:
65ee06171ea45c4dd49aa82c1a04a7beb1501cfe disables the auto sending of the quarterly update

what do we want to do with the rest of the TODOs in the PR?  fold them into the v2 ticket or fix them here?

CuriousLearner commented:
@owocki @thelostone-mc Let me know how we can best gather the data in TODOs.

owocki commented:
@CuriousLearner @mbeacom @thelostone-mc i just pushed up a few commits to resolve the remainder of the TODOs for this..  so if you want to merge this and work on the responsive design issues from there, thats fine with me.

mbeacom commented on app/dashboard/models.py:
```diff
@@ -70,6 +70,13 @@ def exclude_by_status(self, excluded_statuses=None):
 
+    def filter_by_status(self, filtered_status=None):
+        """Filter results with a status matching the provided list."""
+        if filtered_status is None:
```
I'm going to add a check in here on `isinstance()` when `filtered_status is not None` to validate we're receiving a list.

CuriousLearner replied:
What should be returned here if `filtered_status` is not a list?

CuriousLearner replied:
@mbeacom Everything else that you suggested is fixed. This is the last bit that needs to be taken care of. Please let me know :)

mbeacom commented on app/dashboard/models.py:
```diff
@@ -1065,6 +1078,67 @@ def stats(self):
+
+        num_completed_bounties = bounties.filter(idx_status__in=['done']).count()
+        terminal_state_bounties = bounties.filter(idx_status__in=['done', 'expired', 'cancelled']).count()
+        total_bounties = bounties.count()
```
If we're not using this count, we might as well remove it.

mbeacom commented on app/dashboard/models.py:
```diff
@@ -1065,6 +1078,67 @@ def stats(self):
+        ])
+
+        num_completed_bounties = bounties.filter(idx_status__in=['done']).count()
+        terminal_state_bounties = bounties.filter(idx_status__in=['done', 'expired', 'cancelled']).count()
```
Maybe we should define what `terminal_state` statuses are, so we can use the filter anywhere and know it matches this.

mbeacom commented on app/dashboard/models.py:
```diff
@@ -1065,6 +1078,67 @@ def stats(self):
 
+    @property
+    def get_quarterly_stats(self):
+        user_active_in_last_quarter = False
```
We need to get in the habit of adding docstrings that adhere to pep-257 when adding new methods and classes.  This is fine for the moment, but something to consider moving forward.

mbeacom commented on app/marketing/management/commands/send_quarterly_stats.py:
```diff
@@ -0,0 +1,84 @@
+        )
+
+    def handle(self, *args, **options):
+
```
No new line padding on methods.

mbeacom commented on app/marketing/management/commands/send_quarterly_stats.py:
```diff
@@ -0,0 +1,84 @@
+        queryset = queryset.order_by('email')
+        email_list = set(queryset.values_list('email', flat=True))
+
+        print("got {} emails".format(len(email_list)))
```
When possible, please use literal string interpolation when necessary or standard print functionality if type changes aren't necessary:

`print('got {len(email_list)} emails')` or `print('got', len(email_list), 'emails')`

mbeacom commented on app/marketing/management/commands/send_quarterly_stats.py:
```diff
@@ -0,0 +1,84 @@
+        platform_wide_stats = get_platform_wide_stats()
+
+        counter = 0
+        for to_email in email_list:
```
This could be optimized/simplified with `enumerate()`:

`for counter, to_email in enumerate(email_list):`

mbeacom commented on app/marketing/management/commands/send_quarterly_stats.py:
```diff
@@ -0,0 +1,84 @@
+        counter = 0
+        for to_email in email_list:
+            counter += 1
+            print("-sending {} / {}".format(counter, to_email))
```
fstring or standard print usage

mbeacom commented on app/marketing/utils.py:
```diff
@@ -141,3 +142,57 @@ def get_or_save_email_subscriber(email, source, send_slack_invite=True, profile=
+    completed_bounties_fund = round(completed_bounties_fund)
+    bounties_completion_percent = round(bounties_completion_percent)
+
+    largest_bounty = Bounty.objects.filter(current_bounty=True, created_on__gte=last_n_days).order_by('-_val_usd_db').first()
```
This.

mbeacom commented on app/marketing/utils.py:
```diff
@@ -141,3 +142,57 @@ def get_or_save_email_subscriber(email, source, send_slack_invite=True, profile=
+    bounties_completion_percent = round(bounties_completion_percent)
+
+    largest_bounty = Bounty.objects.filter(current_bounty=True, created_on__gte=last_n_days).order_by('-_val_usd_db').first()
+    largest_bounty_value = largest_bounty.value_in_usdt
```
this

mbeacom commented on app/marketing/utils.py:
```diff
@@ -141,3 +142,56 @@ def get_or_save_email_subscriber(email, source, send_slack_invite=True, profile=
+    hunters = [username[0] for username in profiles]
+
+    # Overall transactions across the network are hard-coded for now
+    total_transaction_in_usd = round(sum([bounty.value_in_usdt for bounty in completed_bounties if bounty.value_in_usdt]))
```
^

mbeacom commented on app/marketing/utils.py:
```diff
@@ -141,3 +142,56 @@ def get_or_save_email_subscriber(email, source, send_slack_invite=True, profile=
+
+    # Overall transactions across the network are hard-coded for now
+    total_transaction_in_usd = round(sum([bounty.value_in_usdt for bounty in completed_bounties if bounty.value_in_usdt]))
+    total_transaction_in_eth = round(sum([bounty.value_in_eth for bounty in completed_bounties if bounty.value_in_eth]) / 10**18)
```
^

mbeacom commented on app/retail/templates/emails/quarterly_stats.html:
```diff
@@ -0,0 +1,302 @@
+                <p class="content">{{user_total_earned_usd|intcomma}} USD</p>
+            </div>
+            <div class="stat-contents col-sm-12">
+                <p>You earned {{user_total_earned_eth|intcomma}} ETH ({{user_total_earned_usd|intcomma}} USD) from {{user_fulfilled_bounties_count|intcomma}} completed bounties, for an average of {{user_avg_eth_earned_per_bounty|intcomma}} ETH ({{ user_avg_usd_earned_per_bounty|intcomma}} USD) per bounty!</p>
```
Typically, you want to use space django template variables like: `{{ example }}`, but obviously {{example}} works.  In any case, we should remain consistent, one way or another.

Ref: `({{ user_avg_usd_earned_per_bounty|intcomma}} USD)`

mbeacom commented on app/retail/templates/emails/quarterly_stats.html:
```diff
@@ -0,0 +1,302 @@
+        </div>
+
+        <div class="button-action">
+            <a class="button" href="https://gitcoin.co/explorer">{% trans "Share my stats" %}</a>
```
This should use `{% url 'dashboard' %}`

mbeacom commented on app/retail/templates/emails/quarterly_stats.html:
```diff
@@ -0,0 +1,302 @@
+    <div class="centered-contents stat-box">
+        <div class="stat-img">
+            {% for hunter in hunters %}
+            <img style="border-radius: 50%; width: 35%;" src="https://gitcoin.co/static/avatar/{{hunter}}" />
```
`{% url %}`

mbeacom commented on app/retail/templates/emails/quarterly_stats.html:
```diff
@@ -0,0 +1,302 @@
+    {% if largest_bounty_value %}
+        <div class="centered-contents stat-box">
+            <div class="stat-img">
+                <img style="width: 100%; border-radius: 50%;" src="https://gitcoin.co/static/avatar/{{largest_bounty.bounty_owner_github_username}}" />
```
`{% url %}

mbeacom commented on app/retail/templates/emails/quarterly_stats.html:
```diff
@@ -0,0 +1,302 @@
+        </div>
+    {% endif %}
+    <div class="button-action">
+        <a class="button" href="https://gitcoin.co/explorer">{% trans "Explore New Bounties" %}</a>
```
`{% url %}`

mbeacom commented on app/marketing/utils.py:
```diff
@@ -141,3 +142,57 @@ def get_or_save_email_subscriber(email, source, send_slack_invite=True, profile=
+
+def get_platform_wide_stats(since_last_n_days=90):
+    """
+    get platform wide stats for quarterly stats email
```
pep-257/258 docstrings:

```python
def get_platform_wide_stats(since_last_n_days=90):
    """Get platform wide stats for quarterly stats email."""
```
or
```python
def get_platform_wide_stats(since_last_n_days=90):
    """Get platform wide stats for quarterly stats email.

    Args:
        since_last_n_days (int): The number of days from now to retrieve stats.

    Returns:
        dict: The platform statistics dictionary.

    """
```

PixelantDesign commented:
@CuriousLearner could you remove the tracking on the body copy?

CuriousLearner commented:
@PixelantDesign Sorry, but I didn't get what do you mean by removing tracking. Can you please elaborate a bit?

PixelantDesign commented:
extra spacing between the letters. thanks!

thelostone-mc commented:
@PixelantDesign The letter spacing looks alright to me 🤔 
@CuriousLearner Could you resolve this one last time and we'll get it merged

CuriousLearner commented:
@thelostone-mc Can you please check now?
@PixelantDesign Yeah, I'm not sure about what needs to be done with letter spacing.

owocki commented:
@mbeacom go ahead and merge when youre ready.  unless you want changes

owocki commented:
just funded https://github.com/gitcoinco/web/issues/1179

thelostone-mc commented:
Thanks @CuriousLearner for helping out and sorry for dragging this out 😅
